internal/controllers/configuration/kongadminapi_controller_envtest_test.go

```diff
@@ -0,0 +1,410 @@
+	"time"
+
+	"github.com/bombsimon/logrusr/v2"
+	"github.com/google/uuid"
+	"github.com/samber/lo"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	discoveryv1 "k8s.io/api/discovery/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	ctrl "sigs.k8s.io/controller-runtime"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/adminapi"
+	"github.com/kong/kubernetes-ingress-controller/v2/test/envtest"
+)
+
+// Note: If you want apiserver output on stdout set
+// KUBEBUILDER_ATTACH_CONTROL_PLANE_OUTPUT to true when running tests.
+
+type notifier struct {
+	lock sync.RWMutex
+	t    *testing.T
+	last []adminapi.DiscoveredAdminAPI
+}
+
+func (n *notifier) Notify(adminAPIs []adminapi.DiscoveredAdminAPI) {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+	n.last = adminAPIs
+}
+
+func (n *notifier) LastNotified() []adminapi.DiscoveredAdminAPI {
+	n.lock.RLock()
+	defer n.lock.RUnlock()
+	return n.last
+}
+
+// prepareEnv starts KongAdminAPIServiceReconciler with the manager in a separate
+// goroutine.
+func prepareEnv(ctx context.Context, t *testing.T, client ctrlclient.Client, cfg *rest.Config) (
+	adminService corev1.Service,
+	adminPod corev1.Pod,
+	n *notifier,
+) {
+	ns := envtest.Namespace(ctx, t, client)
+	adminPod = envtest.Pod(ctx, t, client, ns.Name)
+
+	mgr, err := ctrl.NewManager(cfg, ctrl.Options{
+		Logger:             logrusr.New(logrus.New()),
+		Scheme:             scheme.Scheme,
+		SyncPeriod:         lo.ToPtr(2 * time.Second),
+		MetricsBindAddress: "0",
+	})
+	require.NoError(t, err)
+
+	adminService = corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "kong-admin",
+			Namespace: ns.Name,
+			UID:       types.UID(uuid.NewString()),
+		},
+	}
+
+	n = &notifier{t: t}
+	require.NoError(t,
+		(&KongAdminAPIServiceReconciler{
+			Client: mgr.GetClient(),
+			ServiceNN: types.NamespacedName{
+				Name:      adminService.Name,
+				Namespace: adminService.Namespace,
+			},
+			PortNames:         sets.New("admin"),
+			EndpointsNotifier: n,
+			Log:               mgr.GetLogger(),
+		}).SetupWithManager(mgr),
+	)
+	// This wait group makes it so that we wait for manager to exit.
+	// This way we get clean test logs not mixing between tests.
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		assert.NoError(t, mgr.Start(ctx))
+	}()
+	t.Cleanup(func() { wg.Wait() })
+
+	return
+}
+
+func TestKongAdminAPIController(t *testing.T) {
+	t.Parallel()
+
+	// In tests below we use a deferred cancel to stop the manager and not wait
+	// for its timeout.
+
+	cfg := envtest.Setup(t)
+	client, err := ctrlclient.New(cfg, ctrlclient.Options{})
+	require.NoError(t, err)
+
+	t.Run("Endpoints are matched properly", func(t *testing.T) {
```

Comment:
Could we add a test case in which the number of ready endpoints would change over time (let's say 2 -> 0 -> 1) to ensure that notifications are properly propagated in such a scenario?

Reply:
Added.